pkg/kubelet/secret/secret_manager.go

```diff
@@ -73,190 +67,35 @@ func (s *simpleSecretManager) RegisterPod(pod *v1.Pod) {
-	item.secret = nil
-}
-
-func (s *secretStore) Delete(namespace, name string) {
-	key := objectKey{namespace: namespace, name: name}
-
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	if item, ok := s.items[key]; ok {
-		item.refCount--
-		if item.refCount == 0 {
-			delete(s.items, key)
-		}
-	}
-}
-
-func GetObjectTTLFromNodeFunc(getNode func() (*v1.Node, error)) GetObjectTTLFunc {
-	return func() (time.Duration, bool) {
-		node, err := getNode()
-		if err != nil {
-			return time.Duration(0), false
-		}
-		if node != nil && node.Annotations != nil {
-			if value, ok := node.Annotations[v1.ObjectTTLAnnotationKey]; ok {
-				if intValue, err := strconv.Atoi(value); err == nil {
-					return time.Duration(intValue) * time.Second, true
-				}
-			}
-		}
-		return time.Duration(0), false
-	}
-}
-
-func (s *secretStore) isSecretFresh(data *secretData) bool {
-	secretTTL := s.defaultTTL
-	if ttl, ok := s.getTTL(); ok {
-		secretTTL = ttl
-	}
-	return s.clock.Now().Before(data.lastUpdateTime.Add(secretTTL))
-}
-
-func (s *secretStore) Get(namespace, name string) (*v1.Secret, error) {
-	key := objectKey{namespace: namespace, name: name}
-
-	data := func() *secretData {
-		s.lock.Lock()
-		defer s.lock.Unlock()
-		item, exists := s.items[key]
-		if !exists {
-			return nil
-		}
-		if item.secret == nil {
-			item.secret = &secretData{}
-		}
-		return item.secret
-	}()
-	if data == nil {
-		return nil, fmt.Errorf("secret %q/%q not registered", namespace, name)
-	}
-
-	// After updating data in secretStore, lock the data, fetch secret if
-	// needed and return data.
-	data.Lock()
-	defer data.Unlock()
-	if data.err != nil || !s.isSecretFresh(data) {
-		opts := metav1.GetOptions{}
-		if data.secret != nil && data.err == nil {
-			// This is just a periodic refresh of a secret we successfully fetched previously.
-			// In this case, server data from apiserver cache to reduce the load on both
-			// etcd and apiserver (the cache is eventually consistent).
-			util.FromApiserverCache(&opts)
-		}
-		secret, err := s.kubeClient.CoreV1().Secrets(namespace).Get(name, opts)
-		if err != nil && !apierrors.IsNotFound(err) && data.secret == nil && data.err == nil {
-			// Couldn't fetch the latest secret, but there is no cached data to return.
-			// Return the fetch result instead.
-			return secret, err
-		}
-		if (err == nil && !isSecretOlder(secret, data.secret)) || apierrors.IsNotFound(err) {
-			// If the fetch succeeded with a newer version of the secret, or if the
-			// secret could not be found in the apiserver, update the cached data to
-			// reflect the current status.
-			data.secret = secret
-			data.err = err
-			data.lastUpdateTime = s.clock.Now()
-		}
-	}
-	return data.secret, data.err
+// store is the interface for a secrets cache that
+// can be used by cacheBasedSecretManager.
+type store interface {
+	// Add a secret to the store.
+	Add(namespace, name string)
+	// Delete a secret from the store.
```

Comment:
doc refcount impl, that the secret is only removed once every Add() has had a corresponding Delete() called

Reply:
done